### ISSUE TITLE
ONNX conversion of torch model with updating registered buffer

### ISSUE THREAD

**Issue by shivadb** [pytorch/pytorch#60828]
## 🐛 Bug

I have a torch model that has an updating buffer (similar to batchnorm, but I need to update it during inference as well). I am able to reproduce the correct behavior when converting to ScriptModule using torch.jit.script. But when converting to ONNX, the resulting model seems to treat the buffer as a constant and only updates the last value as part of its computation graph. Please see the toy model code below as a simple example:

## To Reproduce

1. Run the following code block:

        import torch
        import torch.nn as nn
        import onnx
        import onnxruntime as rt
        from onnx import helper, shape_inference
        
        print(f'torch version: {torch.__version__}')
        print(f'onnx version: {onnx.__version__}')
        print(f'onnxruntime version: {rt.__version__}')
        
        class InputCache(nn.Module):
            def __init__(self, tensor):
                super(InputCache, self).__init__()
                self.register_buffer('cache', tensor)
            
            def forward(self, x):
                # self.cache[:,:,:-1] = self.cache.clone()[:,:,1:]
                # self.cache[:,:,-1:] = x.detach()
                cache_update = torch.cat((self.cache[:,:,1:], x.detach()), dim=2)
                self.cache = cache_update
                return self.cache
        
        tc = InputCache(torch.zeros(1,1,10))
        tc_script = torch.jit.script(InputCache(torch.zeros(1,1,10)))
        
        # Make sure torch code is doing what it should
        print('Original:')
        for inp in [1.,2.,3.,4.]:
            print(tc(torch.tensor([[[inp]]])))
        
        print()
        print('Convert using script:')
        for inp in [1.,2.,3.,4.]:
            print(tc_script(torch.tensor([[[inp]]])))
        
        # Reinitialize
        tc = InputCache(torch.zeros(1,1,10))
        tc_script = torch.jit.script(InputCache(torch.zeros(1,1,10)))
        
        ex_out = tc(torch.tensor([[[0.5]]]))
        
        print()
        torch.onnx.export(
            tc_script,
            torch.tensor([[[0.0]]]),
            f'onnxrt_test.onnx',
            export_params=True,
            do_constant_folding=True,
            keep_initializers_as_inputs=True,
            opset_version=12,
            input_names = ['input'],
            output_names = ['output'],
            example_outputs=ex_out,
            verbose=True
        )
        
        # Use onnx tools to make sure model is valid
        onnx_model = onnx.load('onnxrt_test.onnx')
        onnx.checker.check_model(onnx_model)
        inferred_model = shape_inference.infer_shapes(onnx_model)
        onnx.checker.check_model(inferred_model)
        
        # Run inference session
        onnx_path = 'onnxrt_test.onnx'
        ort_session = rt.InferenceSession(str(onnx_path))
        
        def to_numpy(tensor):
            return tensor.detach().cpu().numpy() if tensor.requires_grad else tensor.cpu().numpy()
        
        print('ONNX Runtime:')
        for inp in [1.,2.,3.,4.]:
            ort_inputs = {ort_session.get_inputs()[0].name: to_numpy(torch.tensor([[[inp]]]))}
            ort_outs = ort_session.run(None, ort_inputs)
            print(ort_outs[0])

Resulting Output:

        torch version: 1.9.0
        onnx version: 1.9.0
        onnxruntime version: 1.7.2
        Original:
        tensor([[[0., 0., 0., 0., 0., 0., 0., 0., 0., 1.]]])
        tensor([[[0., 0., 0., 0., 0., 0., 0., 0., 1., 2.]]])
        tensor([[[0., 0., 0., 0., 0., 0., 0., 1., 2., 3.]]])
        tensor([[[0., 0., 0., 0., 0., 0., 1., 2., 3., 4.]]])
        
        Convert using script:
        tensor([[[0., 0., 0., 0., 0., 0., 0., 0., 0., 1.]]])
        tensor([[[0., 0., 0., 0., 0., 0., 0., 0., 1., 2.]]])
        tensor([[[0., 0., 0., 0., 0., 0., 0., 1., 2., 3.]]])
        tensor([[[0., 0., 0., 0., 0., 0., 1., 2., 3., 4.]]])
        
        graph(%input : Float(1, 1, 1, strides=[1, 1, 1], requires_grad=0, device=cpu),
              %8 : Float(1, 1, 9, strides=[10, 10, 1], requires_grad=0, device=cpu)):
          %output : Float(1, 1, 10, strides=[10, 10, 1], requires_grad=0, device=cpu) = onnx::Concat[axis=2](%8, %input)
          return (%output)
        
        ONNX Runtime:
        [[[0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]]]
        [[[0. 0. 0. 0. 0. 0. 0. 0. 0. 2.]]]
        [[[0. 0. 0. 0. 0. 0. 0. 0. 0. 3.]]]
        [[[0. 0. 0. 0. 0. 0. 0. 0. 0. 4.]]]

## Expected behavior

The ONNX model should be able to reproduce the original PyTorch behaviour. The ONNX Runtime output above should be:

        ONNX Runtime:
        [[[0., 0., 0., 0., 0., 0., 0., 0., 0., 1.]]]
        [[[0., 0., 0., 0., 0., 0., 0., 0., 1., 2.]]]
        [[[0., 0., 0., 0., 0., 0., 0., 1., 2., 3.]]]
        [[[0., 0., 0., 0., 0., 0., 1., 2., 3., 4.]]]

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0): 1.9.0
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source): 
 - Python version: 3.7
 - CUDA/cuDNN version: 10.2
 - GPU models and configuration:
 - Any other relevant information:

## Additional context

I also tried replacing the update operation to force in-place updates the the tensor. This results in a longer and more complicated ONNX graph, but the behavior is the same. Also, changing the registered buffer to a registered parameter, or just a Python variable results in the same behavior.

cc @garymm @BowenBao @neginraoof

**Comment by shivadb** [comment]
Any advice on this? I still have not been able to find a solution/workaround.

**Comment by BowenBao** [comment]
hi @shivadb, this is fundamentally not possible for ORT inference since the ONNX model is stateless. A suggested workaround is to change `cache` to input/output argument in `forward`, and manage it outside of ORT call.

**Comment by garymm** [comment]
@BowenBao I wonder if it would be better to fail in export when we detect state-modifying behavior like this rather than export it and just silently drop the state modification. Has this been discussed?

**Comment by shivadb** [comment]
Thanks for clarifying @BowenBao. I was afraid that was the case. I am actually using this type of module in many places in the network so I was trying to avoid having to separate the network into many ORT calls. Sounds like that can't be avoided.

Thanks for you help.
